# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
#load python packages
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
#print out current directory
os.getcwd()

'C:\\Users\\thoma\\projects\\datascience\\capstone-01\\notebooks'

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [2]:
path = 'C:\\Users\\thoma\\projects\\datascience\\capstone-01\\data\\processed'
print(os.listdir(path))
os.chdir(path)

df = pd.read_csv('step3_output.csv',)
df.head()

['step2_output.csv', 'step3_output.csv']


Name    state  summit_elev  vertical_drop  trams  fastEight  \
0       Alyeska Resort   Alaska         3939           2500      1          0   
1  Eaglecrest Ski Area   Alaska         2600           1540      0          0   
2     Hilltop Ski Area   Alaska         2090            294      0          0   
3     Arizona Snowbowl  Arizona        11500           2300      0          0   
4  Sunrise Park Resort  Arizona        11100           1800      0          0   

   fastSixes  fastQuads  quad  triple  ...  SkiableTerrain_ac  Snow Making_ac  \
0          0          2     2       0  ...             1610.0           113.0   
1          0          0     0       0  ...              640.0            60.0   
2          0          0     0       1  ...               30.0            30.0   
3          1          0     2       2  ...              777.0           104.0   
4          0          1     2       3  ...              800.0            80.0   

   daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  AdultWeekend  \
0               150         60            669.0          65.0          85.0   
1                45         44            350.0          47.0          53.0   
2               150         36             69.0          30.0          34.0   
3               122         81            260.0          89.0          89.0   
4               115         49            250.0          74.0          78.0   

   projectedDaysOpen  NightSkiing_ac  clusters  
0                150           550.0         0  
1                 90             0.0         0  
2                152            30.0         0  
3                122             0.0         1  
4                104            80.0         1  

[5 rows x 26 columns]

## Drop weekday prices
Hight correlation between weekday and weekend prices, and both can be adjusted independently, so weekday prices will be removed

In [3]:
df.drop('AdultWeekday', axis=1, inplace=True)

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [4]:
dfo=df[['state']] # select object type columns and then deselect name
print(dfo.head())

df1 = pd.concat([df.drop(dfo, axis=1), pd.get_dummies(dfo)], axis=1)

print(df1.info())

     state
0   Alaska
1   Alaska
2   Alaska
3  Arizona
4  Arizona
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 330 entries, 0 to 329
Data columns (total 59 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Name                  330 non-null    object 
 1   summit_elev           330 non-null    int64  
 2   vertical_drop         330 non-null    int64  
 3   trams                 330 non-null    int64  
 4   fastEight             330 non-null    int64  
 5   fastSixes             330 non-null    int64  
 6   fastQuads             330 non-null    int64  
 7   quad                  330 non-null    int64  
 8   triple                330 non-null    int64  
 9   double                330 non-null    int64  
 10  surface               330 non-null    int64  
 11  total_chairs          330 non-null    int64  
 12  Runs                  330 non-null    int64  
 13  TerrainParks          330 non-null    int64  
 14  LongestR

In [5]:
df2 = df.drop('state',axis=1)
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 330 entries, 0 to 329
Data columns (total 24 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Name               330 non-null    object 
 1   summit_elev        330 non-null    int64  
 2   vertical_drop      330 non-null    int64  
 3   trams              330 non-null    int64  
 4   fastEight          330 non-null    int64  
 5   fastSixes          330 non-null    int64  
 6   fastQuads          330 non-null    int64  
 7   quad               330 non-null    int64  
 8   triple             330 non-null    int64  
 9   double             330 non-null    int64  
 10  surface            330 non-null    int64  
 11  total_chairs       330 non-null    int64  
 12  Runs               330 non-null    int64  
 13  TerrainParks       330 non-null    int64  
 14  LongestRun_mi      330 non-null    float64
 15  SkiableTerrain_ac  330 non-null    float64
 16  Snow Making_ac     330 non

In [6]:
df3 = df.drop(['state','summit_elev','vertical_drop','averageSnowfall','yearsOpen','daysOpenLastYear','projectedDaysOpen'], axis=1)
df3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 330 entries, 0 to 329
Data columns (total 18 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Name               330 non-null    object 
 1   trams              330 non-null    int64  
 2   fastEight          330 non-null    int64  
 3   fastSixes          330 non-null    int64  
 4   fastQuads          330 non-null    int64  
 5   quad               330 non-null    int64  
 6   triple             330 non-null    int64  
 7   double             330 non-null    int64  
 8   surface            330 non-null    int64  
 9   total_chairs       330 non-null    int64  
 10  Runs               330 non-null    int64  
 11  TerrainParks       330 non-null    int64  
 12  LongestRun_mi      330 non-null    float64
 13  SkiableTerrain_ac  330 non-null    float64
 14  Snow Making_ac     330 non-null    float64
 15  AdultWeekend       330 non-null    float64
 16  NightSkiing_ac     330 non

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [7]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X1 = df1.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.loc[:,'AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X1)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled1 = scaler.transform(X1) 
X_scaled1

array([[-0.1750243 ,  1.3572862 ,  1.4796585 , ..., -0.11076976,
        -0.22573306, -0.15762208],
       [-0.53401788,  0.34294523, -0.30893969, ..., -0.11076976,
        -0.22573306, -0.15762208],
       [-0.6707518 , -0.97358483, -0.30893969, ..., -0.11076976,
        -0.22573306, -0.15762208],
       ...,
       [ 0.86227679,  0.37569999, -0.30893969, ..., -0.11076976,
        -0.22573306,  6.34428877],
       [ 1.35961292, -0.23818762, -0.30893969, ..., -0.11076976,
        -0.22573306,  6.34428877],
       [ 1.31591169, -0.12196105, -0.30893969, ..., -0.11076976,
        -0.22573306,  6.34428877]])

In [8]:
# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X2 = df2.drop(['Name','AdultWeekend'], axis=1)

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X2)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled2 = scaler.transform(X2) 
X_scaled2

array([[-1.75024298e-01,  1.35728620e+00,  1.47965850e+00, ...,
         1.04373424e+00,  5.28388759e+00, -7.36244575e-01],
       [-5.34017876e-01,  3.42945226e-01, -3.08939686e-01, ...,
        -1.04690187e+00, -6.09611897e-01, -7.36244575e-01],
       [-6.70751800e-01, -9.73584832e-01, -3.08939686e-01, ...,
         1.11342211e+00, -2.88148289e-01, -7.36244575e-01],
       ...,
       [ 8.62276787e-01,  3.75699987e-01, -3.08939686e-01, ...,
         1.02947990e-01,  5.69088000e-01,  1.35824430e+00],
       [ 1.35961292e+00, -2.38187625e-01, -3.08939686e-01, ...,
        -1.58381523e-03, -6.09611897e-01,  1.35824430e+00],
       [ 1.31591169e+00, -1.21961055e-01, -3.08939686e-01, ...,
        -1.58381523e-03, -6.09611897e-01,  1.35824430e+00]])

In [9]:
# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X3 = df3.drop(['Name','AdultWeekend'], axis=1)

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X3)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled3 = scaler.transform(X3) 
X_scaled3

array([[ 1.4796585 , -0.05513178, -0.2840776 , ..., -0.1503032 ,
         5.28388759, -0.73624458],
       [-0.30893969, -0.05513178, -0.2840776 , ..., -0.36274889,
        -0.6096119 , -0.73624458],
       [-0.30893969, -0.05513178, -0.2840776 , ..., -0.48300117,
        -0.28814829, -0.73624458],
       ...,
       [-0.30893969, -0.05513178, -0.2840776 , ...,  0.39884887,
         0.569088  ,  1.3582443 ],
       [-0.30893969, -0.05513178, -0.2840776 , ..., -0.48300117,
        -0.6096119 ,  1.3582443 ],
       [-0.30893969, -0.05513178, -0.2840776 , ..., -0.60325344,
        -0.6096119 ,  1.3582443 ]])

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [10]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

In [11]:
# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train1, X_test1, y_train1, y_test1 = train_test_split(X_scaled1, y, test_size=0.25, random_state=1)

In [12]:
# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train2, X_test2, y_train2, y_test2 = train_test_split(X_scaled2, y, test_size=0.25, random_state=1)

In [13]:
# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train3, X_test3, y_train3, y_test3 = train_test_split(X_scaled3, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Multiple Linear Regression Formula
The Formula for Multiple Linear Regression is:
\begin{aligned} &y_i = \beta_0 + \beta _1 x_{i1} + \beta _2 x_{i2} + ... + \beta _p x_{ip} + \epsilon\\ &\textbf{where, for } i = n \textbf{ observations:}\\ &y_i=\text{dependent variable}\\ &x_i=\text{expanatory variables}\\ &\beta_0=\text{y-intercept (constant term)}\\ &\beta_p=\text{slope coefficients for each explanatory variable}\\ &\epsilon=\text{the model's error term (also known as the residuals)}\\ \end{aligned} 
[SOURCE](https://www.investopedia.com/terms/m/mlr.asp)

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [14]:
# all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error

lm = linear_model.LinearRegression()
model = lm.fit(X_train1,y_train1)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [15]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred1 = model.predict(X_test1)

## Review Model Outcomes — Iterate over additional models as needed

In [16]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# You can plug y_test and y_pred into the functions to evaluate the model
explained_variance_score(y_test1, y_pred1)

0.3662666865074793

In [17]:
mean_absolute_error(y_test1, y_pred1)

13.763082588095843

**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [18]:
lm.intercept_

64.27769116948181

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

## NOTE
It was after running this code snippet that I realized dropping a large number of observations in the data cleaning phase was negatively affecting my model. The fastEight and trams variables were having a profound effect on the model most likely because these variables became binary due to the large number of "outliers" removed for having more than 1 of them. Data was recleaned after this point, increasing the number of observations from 175 to 330

In [19]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X1.columns, columns=['Coefficient']).sort_values('Coefficient', ascending=False).head(10)


Coefficient
state_New York       3.168573e+13
state_Michigan       2.990272e+13
state_Colorado       2.634603e+13
state_California     2.578205e+13
state_Pennsylvania   2.460285e+13
state_New Hampshire  2.268576e+13
state_Wisconsin      2.268576e+13
state_Vermont        2.200034e+13
state_Minnesota      2.128806e+13
state_Utah           2.054612e+13

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [75]:
lm = linear_model.LinearRegression()
model = lm.fit(X_train2,y_train2)

In [76]:
y_pred2 = model.predict(X_test2)

In [77]:
explained_variance_score(y_test2, y_pred2)

0.2738040416288373

In [78]:
mean_absolute_error(y_test2, y_pred2)

14.750532371455671

In [79]:
lm.intercept_

64.24404947702106

In [80]:
pd.DataFrame(abs(lm.coef_), X2.columns, columns=['Coefficient']).sort_values('Coefficient', ascending=False).head(10)

Coefficient
summit_elev        8.989751
clusters           8.926219
Snow Making_ac     8.308429
vertical_drop      5.798553
fastQuads          2.921294
triple             2.876033
surface            2.660582
quad               2.632999
trams              2.267814
fastSixes          1.919110

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [81]:
lm = linear_model.LinearRegression()
model = lm.fit(X_train3,y_train3)

In [82]:
y_pred3 = model.predict(X_test3)

In [83]:
explained_variance_score(y_test3, y_pred3)

0.2919119878081584

In [84]:
mean_absolute_error(y_test3, y_pred3)

15.081379005019729

In [85]:
lm.intercept_

64.14753298816537

In [86]:
pd.DataFrame(abs(lm.coef_), X3.columns, columns=['Coefficient']).sort_values('Coefficient', ascending=False).head(10)

Coefficient
Snow Making_ac     8.491579
Runs               5.060119
fastQuads          3.783783
quad               3.329143
surface            3.308062
fastSixes          3.186773
trams              2.996485
fastEight          2.603331
triple             2.551908
LongestRun_mi      2.127807

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | 0.3663 | 13.763 |-|
| Model 2. | 0.2738 | 14.751 |'state'|
| Model 3. | 0.2919 | 15.081 |'state','summit_elev','base_elev','averageSnowfall','yearsOpen','daysOpenLastYear','projectedDaysOpen'|

**Model Selection:** Model 3 - It removes inherent features that the resort cannot control. Note though that when comparing to the first model, it has both a lower explained variance and a higher mean absolute error, meaning it is less accurate on purely a numbers perspective.